### PR TITLE
Bohandley/grafana assistant traces drilldown full query navigation

### PR DESCRIPTION
**Pull request by bohandley**
Fixes https://github.com/grafana/grafana-assistant-app/issues/959

### What is this?

This is a Grafana Assistant integration. This is to use the traces drilldown links.ts to generate a correct URL for the Grafana assistant to navigate to the traces drilldown app.

This is part of this [epic here to standardize the drilldown routes in the Grafana assistant](https://github.com/grafana/grafana-assistant-app/issues/956).

### What changes are in the code?

1. The `contextToLink` function in links.ts now has the following features
  - can return a root URL for traces drilldown app
  - can parse a limited query (similar to Tempo data source)
2. Traces drilldown adds `grafana-assistant-app/navigateToDrilldown/v1` extension point in the linksConfig targets.

### Why a limited query parser?

I completed this analysis with two strategies 1) limited parser & 2) full parser using lezer traceql

##### Two Complete Implementations

1. Limited Regex Parser ⭐ RECOMMENDED
  - limited-traceql-parser.ts - Lightweight, focused parser
  - Enhanced links.ts with fallback to raw query parsing
  - Bundle impact: +6KB (+50% from baseline)
2. Full Lezer Parser
  - full-traceql-parser.ts - Complete AST-based parser
  - links-with-full-parser.ts - Alternative implementation
  - Bundle impact: +83KB (+692% from baseline)

##### 📊 Comprehensive Analysis Reports
  - [bundle-analysis-report.txt](https://github.com/user-attachments/files/21765470/bundle-analysis-report.txt) - 12KB detailed technical report
  - [bundle-comparison.csv](https://github.com/user-attachments/files/21765472/bundle-comparison.csv) - Comparison spreadsheet for stakeholders
  - 
##### 🏆 Key Findings & Recommendation
  - The Limited Parser approach is strongly recommended:
  - Bundle Size: 1,283% smaller impact than full parser (6KB vs 83KB)
  - Functionality: Covers 95%+ of expected TraceQL patterns for links.ts
  - Performance: Minimal impact on initial page load (+6ms vs +83ms)
  - Maintenance: Simple, focused implementation without heavy dependencies

The analysis shows that the limited parser provides the optimal balance of functionality and performance for enabling raw TraceQL query support in the traces-drilldown plugin's contextToLink function.All files are ready for integration and the bundle entry point size impact has been minimized while providing the essential functionality needed for the traces-drilldown app extension.

### CODE REVIEW

File: src/utils/limited-traceql-parser.ts [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+ * Examples:
+ * - {resource.service.name="my-service"}
+ * - {status=error}
+ * - {span.http.status_code=200 && resource.service.name="api"}
+ * - {span:duration>100ms}
+ * 
+ * @param query - Raw TraceQL query string
+ * @returns Array of TraceqlFilter objects or null if parsing fails
+ */
+export function parseTraceQLQuery(query: string): TraceqlFilter[] | null {
```
[the remark this answers is not in the document]
Comment (ifrost) [reply src/utils/limited-traceql-parser.ts:43]: > @ifrost what are your thoughts on this PR implementing the lezer-traceql parser? It's main drawback is bundle entry point size increase. Generally we want to keep this below ~250kB but does traces-drilldown have a lower target size? I would be happy to make the changes for that in this PR if you are ok with it. What do you think?

I don't think there's any hard limit for Traces Drilldown at the moment 🤔  (@joey-grafana?) Sounds reasonable to use lezer-traceql parser!

[the remark this answers is not in the document]
Comment (bohandley) [reply src/utils/limited-traceql-parser.ts:43]: Let's go lezer parser traceql!!! 🤘

**Comment by joey-grafana** [comment]
@bohandley I'm curious where did the requirement to navigate to Traces Drilldown with a query come from? I don't remember planning this feature inside the team or in any of our crossover meetings.

**Comment by bohandley** [comment]
> @bohandley I'm curious where did the requirement to navigate to Traces Drilldown with a query come from? I don't remember planning this feature inside the team or in any of our crossover meetings.

@joey-grafana Good point! This makes sense that @grafana/observability-traces-and-profiling would have a standard for introducing new work. If there is anything I can do moving forward to align with this, I am happy to do so.

This work comes from a [request](https://raintank-corp.slack.com/archives/C075BDBTX96/p1752691176094509) of @Rperry2174 to standardize the navigation of the drilldown apps in Grafana Assistant. This thread has a lot going on in it but the gist is that Grafana Assistant wants simple context and consistent url creation across the apps. There previous strategy was resulting in broken links and they want this working in place for ObscCON. The previous strategy for the Grafana assistant drilldown app navigation was custom reverse engineering of drilldown URLs. Each drilldown app required the LLM to learn a lot of specific, custom things for each URL. @graph-andrew suggested I begin working on this and [I created an issue with some rationale for choosing the query strategy](https://github.com/grafana/grafana-assistant-app/issues/956) (I believe now I needed a design document, which I'm making this week). I am sorry for this communication going on in a way that didn't get surfaced to your squad.

Let me know if you'd like to schedule a meeting with Ryan or Andrew for more context, or if there is something I can do better to share more info on this work!

**Comment by joey-grafana** [comment]
Thanks for the added context. I had seen the Slack conversation but was not sure how we came about deciding the current approach as the best path forward.

I'm essentially looking for use cases to understand better where the parsing should happen. Looks like you may have been using writing the query in the Assistant to stimulate a link just to see if we would go to the correct place and run the correct query but it's mostly the button links themselves (from the Assistant) that you want to make sure navigate correctly?

Having thought about it some more, I feel like the parsing would indeed be better off using TraceQL Lezer, what do you think? The benefit here is we could then share the code between repos and it would also align with what Metrics and Logs are doing.

**Comment by ifrost** [comment]
> Having thought about it some more, I feel like the parsing would indeed be better off using TraceQL Lezer...

If I understand correctly there are various use cases on how Assistant need to work with a query:
- (1) It can get a raw query from a Dashboard and need to open it in Explore / Traces Drilldown (I think there's no need to parse it, just pass it along)
- (2) The user may provide service name and ask to query data fro that service (Drilldown apps should probably have capability to handle it to some extend, so there's no need to build a query)
- (3) The user may ask to explain a query (not sure if this needs parsing or LLM can handle it?)

I wonder if Assistant really needs to have parsing capabilities built in or could treat a raw query more like a black box that is passed as input 🤔

**Comment by bohandley** [comment]
> If I understand correctly there are various use cases on how Assistant need to work with a query:
> 
> (1) It can get a raw query from a Dashboard and need to open it in Explore / Traces Drilldown (I think there's no need to parse it, just pass it along)
> (2) The user may provide service name and ask to query data fro that service (Drilldown apps should probably have capability to handle it to some extend, so there's no need to build a query)
> (3) The user may ask to explain a query (not sure if this needs parsing or LLM can handle it?)

Yes, @ifrost, the query may come from an any context dashboard, panel, alert rule, user generated. A user may not even have a query but have a "concept" of a query, i.e., a service. 

> I wonder if Assistant really needs to have parsing capabilities built in or could treat a raw query more like a black box that is passed as input 🤔

Yes! This is what I'm going for. Grafana Assistant should not have to parse a full query. It should, to some extent, know how to build a query from inexact user input. Once it has the query, then we can pass that query along to the drilldown apps and BOOM! 🎇  a correct url is created ✨ The assistant knows just enough and the navigation works nicely. 😄

**Comment by bohandley** [comment]
> Would that 50kB need to be loaded during onClick and possibly making the interaction more laggy? 🤔

@ifrost Not laggy from my testing :). I've been testing the performance in chrome tools and the onclick takes about ~10ms long with the lazy load, so not a noticeable. Here is where the module is loaded and takes about 13ms 💨 

Screenshot 2025-09-03 at 11 37 11 AM

> Is it 50kB gzipped? How does it change the overall module size in %?

I believe it is 50kB gzipped and the overall module size grew 5kB from our implementation.

Lazy loaded in the async onClick callback:
Screenshot 2025-09-03 at 11 47 50 AM

Imported directly, used in the synchronous configure callback:
Screenshot 2025-09-03 at 11 47 42 AM

**Comment by bohandley** [comment]
@ifrost and @joey-grafana 

Here is the lezer parser implementation as well as AHAs, gotchas and extra things I found in implementing this. I am now getting a better picture of query parsing in Tempo! From query builder to raw query makes sense. From raw query to filters, that the links.ts config function understands to build the url, that is interesting!

# Lezer TraceQL Parser Implementation

## Overview

This PR replaces the regex-based `limited-traceql-parser` with a robust Lezer-based parser for TraceQL query parsing in the traces-drilldown plugin. The new parser uses the official `@grafana/lezer-traceql` package for accurate syntax tree parsing and leverages helper functions from the Tempo datasource for consistency.

## Why Use the Lezer Parser?

### Benefits
- **🎯 100% Syntax Accuracy**: Eliminates regex edge cases and false positives
- **🔄 Future-Proof**: Automatically supports new TraceQL syntax as the grammar evolves
- **🔗 Consistency**: Uses the same parser as Grafana's Tempo datasource
- **🛡️ Better Error Handling**: Gracefully handles malformed or complex queries
- **🧪 Testability**: Easier to debug and validate parsing logic

### Costs
- **📦 Bundle Size**: Adds ~50-100KB from `@grafana/lezer-traceql` dependency
- **⚡ Performance**: Slightly more overhead than regex for simple queries
- **🔧 Complexity**: More sophisticated implementation than regex matching

## Parser Scope & Limitations

### ✅ Supported TraceQL Features

The parser supports all TraceQL features that the **Tempo Query Builder** can generate:

- **Basic spanset filters**: `{field=value}`, `{scope.field=value}`
- **Intrinsic fields**: `{duration>100ms}`, `{status=error}`
- **Scoped attributes**: `{resource.service.name="api"}`, `{span.http.status_code=200}`
- **All comparison operators**: `=`, `!=`, `>`, `<`, `>=`, `<=`, `=~`, `!~`
- **Logical AND**: `{field1=value1 && field2=value2}`
- **Multiple spansets**: `{spanset1} && {spanset2}`
- **Multiple values per field**: `{(field="val1" || field="val2")}` → Combined into array values
- **Comments**: `// line comments` and `/* block comments */`

### ❌ Not Supported (Valid TraceQL but Beyond Query Builder Scope)

These are valid TraceQL features but are **not supported by the Tempo Query Builder UI**, so they're outside our parsing scope:

- **OR between different spansets**: `{service="A"} || {service="B"}`
  - *Reason*: Query Builder only creates individual `TraceqlFilter` objects combined with AND logic
- **Structural operators**: `{parent} > {child}`, `{ancestor} >> {descendant}`
  - *Reason*: No Query Builder UI for span relationships
- **Pipeline aggregations**: `{filters} | count() > 5`
  - *Reason*: Query Builder focuses on filtering, not aggregation
- **Union operators**: `&>`, `&>>`, `&<<`, `&<`, `&~`
  - *Reason*: No Query Builder support for complex relationships

## Key Technical Implementation

### Filter Combination Logic

The parser includes a critical feature to handle the Tempo Query Builder's multiple-value pattern:

**Query Builder generates** (`SearchTraceQLEditor/utils.ts` lines 83-84):
```typescript
if (Array.isArray(f.value) && f.value.length > 1 && !isRegExpOperator(f.operator!)) {
  return `(${f.value.map((v) => `${scopeHelper(f, lp)}${tagHelper(f, filters)}${f.operator}${valueHelper({ ...f, value: v })}`).join(' || ')})`;
}
```

**Parser recreates original structure**:
```typescript
// Input: {(resource.service.name="quoteservice" || resource.service.name="cartservice")}
// Output: Single filter with array value: ["quoteservice", "cartservice"]
```

This ensures proper URL encoding as: `var-filters=resource.service.name|=|quoteservice,cartservice`

## Issues Discovered During Investigation

### 1. Variable Filters Display Issue

**Found**: While the parser correctly generates `var-filters` URL parameters with array values (`val1,val2`), these **do not display correctly** in the traces-drilldown variable filters UI.

**Impact**: Users see only one value selected instead of multiple values when navigating from Tempo Query Builder to traces-drilldown.

**Status**: Noted for future investigation - this is a separate issue in the traces-drilldown app's variable handling, not in the link generation.

### 2. TraceQL Syntax Validation

**Found**: Some existing tests were using **invalid TraceQL syntax** that doesn't match the official specification:
- ❌ `{status=error || status=timeout}` (OR within single spanset)
- ✅ `{status=error} || {status=timeout}` (OR between spansets)

**Action**: Updated tests to use correct TraceQL syntax based on official documentation.

## Files Changed

- **Added**: `src/utils/lezer-traceql-parser.ts` - New Lezer-based parser
- **Added**: `src/utils/lezer-traceql-parser.test.ts` - Comprehensive test suite
- **Modified**: `src/utils/links.ts` - Updated import to use new parser
- **Modified**: `package.json` - Added `@grafana/lezer-traceql` dependency

## Testing

- **54 tests passing** across both link generation and parser functionality
- **Comprehensive coverage** of all supported TraceQL patterns
- **Edge case handling** for complex queries and error conditions
- **Query Builder compatibility** tests for real-world usage patterns

**Comment by ifrost** [comment]
> @ifrost Not laggy from my testing :). I've been testing the performance in chrome tools and the onclick takes about ~10ms long with the lazy load, so not a visually noticeable difference. Here is where the module is loaded and takes about 13ms 💨

If it's 10ms I wound't bother about increasing the bundle size and I would keep things simple instead.

**Comment by bohandley** [comment]
@ifrost @svennergr @leeoniya 

For the additional 50kB, I have a working [PR as a performance improvement for Metrics Drilldown](https://github.com/grafana/metrics-drilldown/pull/665). I think this is important but I also think it is secondary to this work of having accurate parsing for url building. Can this be done in a separate PR? to track that we are reducing the bundle entry point size for performance? Or is this blocking and needs to be addressed before moving on with the query parser? What are your thoughts?

File: src/utils/lezer-traceql-parser.ts [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+  'duration', 'kind', 'name', 'status', 'statusMessage', 'traceDuration',
+  'rootName', 'rootServiceName',
+  'event:name', 'event:timeSinceStart',
+  'instrumentation:name', 'instrumentation:version',
+  'link:spanID', 'link:traceID',
+  'span:duration', 'span:id', 'span:kind', 'span:name', 'span:parentID', 
+  'span:status', 'span:statusMessage',
+  'trace:duration', 'trace:id', 'trace:rootName', 'trace:rootService'
+];
+
```
[the remark this answers is not in the document]
Comment (bohandley) [reply src/utils/lezer-traceql-parser.ts]: THis makes sense, but I think that we are looking for a best effort for this PR. I will write an issue to follow up, "Load Valid Scopes and Intrinsic fields from Tempo datasource."

File: src/utils/lezer-traceql-parser.ts [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+/**
+ * Extract filters from a raw TraceQL query string using the Lezer parser.
+ * This provides more accurate parsing than regex-based approaches.
+ * 
+ * @param query - Raw TraceQL query string
+ * @returns Object with filters array and errors array, or null if parsing fails completely
+ */
+export function parseTraceQLQuery(query: string): TraceQLParseResult | null {
```
[the remark this answers is not in the document]
Comment (bohandley) [reply src/utils/lezer-traceql-parser.ts]: We can keep this in traces drilldown!

**Comment by bohandley** [comment]
> @bohandley Thanks for making adjustments for multiple spansets! 🙌 Could you also rebase / resolve conflicts in your branch?

@ifrost all conflicts resolved! Thank you!
